docs/designs/termination.md

```diff
@@ -0,0 +1,130 @@
+
+### Parallel Node Termination
+
+The termination controller will be able to drain and terminate multiple nodes in parallel. We will expedite this process while rate limiting node termination to protect against throttling.
+
+We introduce an optional cluster-scoped CRD, the Node Disruption Budget (NDB), a [Pod Disruption Budget (PDB)](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) for nodes. A user can scope an NDB to Karpenter provisioned nodes through a label selector, since nodes are created with their Provisioner’s labels. `Unavailable` nodes will be `NotReady` or have `metadata.DeletionTimestamp` set. `Available` nodes will be `Ready`.
+
+A termination is allowed if at least minAvailable nodes selected by selector will still be available after the termination. For example, you can prevent all terminations by specifying “100%”. A termination is also allowed if at most maxUnavailable nodes selected by selector are unavailable after the termination. For example, one can prevent all terminations by specifying 0. These two are mutually exclusive.  
+
+Note that this is an experimental idea, and will require robustness improvements for future features such as defragmentation, over-provisioning, and more.  [PodDisruptionBudget Spec](https://pkg.go.dev/k8s.io/api/policy/v1beta1#PodDisruptionBudgetSpec) for reference.
+
+```{go}
+// .go
+type NodeDisruptionBudget struct {
+   // +optional
+   MinAvailable   *intstr.IntOrString   `json:"minAvailable,omitempty" protobuf:"bytes,1,opt,name=minAvailable"`
+   // Node Selector
+   // Label query over nodes managed by the Disruption Budget
+   // A null selector selects no nodes.
+   // An empty selector ({}) also selects no nodes, which differs from standard behavior of selecting all nodes.
+   // +patchStrategy=replace
+   // +optional
+   Selector       *metav1.LabelSelector `json:"selector,omitempty" patchStrategy:"replace" protobuf:"bytes,2,opt,name=selector"`
+   // +optional
+   MaxUnavailable *intstr.IntOrString   `json:"maxUnavailable,omitempty" protobuf:"bytes,1,opt,name=minAvailable"`
+}
+```
+
+```{yaml}
+# .yaml
+apiVersion: termination.karpenter.sh/v1alpha1
+kind: NodeDisruptionBudget
+metadata:
+  name: nodeBudget
+spec:
+  minAvailable: 80%
+  selector:
+    matchLabels:
+      karpenter.sh/name: default
+      karpenter.sh/namespace: default
+  maxUnavailable: 5%
+```
+
+Since most eviction failures come from a PDB violation, we will queue up evictions serially per PDB, where the PDB’s selector will act as the key. The pods will kick off eviction serially, run asynchronously, and exponentially back off and retry if they fail. This way, evicting pods will only be limited by other evicting pods managed by the same PDB, since each pod can only belong to one PDB. Pods that have a `do-not-evict` label will not be queued up for eviction, as we prevent the node with a `do-not-evict` pod from draining.
```

Comment:
What does "The pods will kick off eviction serially"? Do you mean the controller will evict pods serially?

Reply:
I changed it to "The controller will call evictions serially". WDYT?